File: chaincode/sensorchain/src/anchor.ts
/*
 * Merkle-Root Batch Anchor contract (production chaincode).
 * Mirrors sensorchain/contracts/anchor.py.
 *
 * The gateway signs `gatewayId|batchId|merkleRoot|windowStart|windowEnd|count`
 * with its HSM key (ECDSA P-256 / SHA-256, DER signature). The contract
 * verifies the signature against the public key registered at procurement
 * before accepting the anchor.
 */

import { createVerify } from "crypto";
import { Context, Contract, Info, Returns, Transaction } from "fabric-contract-api";
import { deviceKey, DeviceRecord } from "./deviceRegistry";

export interface BatchAnchor {
  batch_id: string;
  gateway_id: string;
  merkle_root: string;
  window_start: number;
  window_end: number;
  reading_count: number;
  signature: string;
  anchored_at: number;
}

export function signingPayload(
  gatewayId: string, batchId: string, merkleRoot: string,
  windowStart: number, windowEnd: number, count: number,
): Buffer {
  return Buffer.from(
    `${gatewayId}|${batchId}|${merkleRoot}|${windowStart.toFixed(3)}|${windowEnd.toFixed(3)}|${count}`,
    "utf8",
  );
}

@Info({
  title: "Anchor",
  description: "Merkle-root batch anchoring with gateway signature verification",
})
export class AnchorContract extends Contract {
  constructor() {
    super("Anchor");
  }

  private key(ctx: Context, batchId: string): string {
    return ctx.stub.createCompositeKey("anchor", [batchId]);
  }

  @Transaction()
  @Returns("string")
  public async AnchorBatch(
    ctx: Context, gatewayId: string, batchId: string, merkleRoot: string,
    windowStart: string, windowEnd: string, count: string, signatureHex: string,
  ): Promise<string> {
    const gatewayBytes = await ctx.stub.getState(deviceKey(ctx, gatewayId));
    if (gatewayBytes.length === 0) {
      throw new Error(`anchor rejected: gateway ${gatewayId} is not registered`);
    }
    const gateway = JSON.parse(gatewayBytes.toString()) as DeviceRecord;
    if (gateway.status === "retired") {
      throw new Error(`anchor rejected: gateway ${gatewayId} is retired`);
    }

    const payload = signingPayload(
      gatewayId, batchId, merkleRoot,
      parseFloat(windowStart), parseFloat(windowEnd), parseInt(count, 10),
    );
    const verifier = createVerify("SHA256");
    verifier.update(payload);
    const valid = verifier.verify(gateway.public_key_pem, Buffer.from(signatureHex, "hex"));
    if (!valid) {
      ctx.stub.setEvent("AnchorSignatureInvalid", Buffer.from(JSON.stringify({
        gateway_id: gatewayId, batch_id: batchId, merkle_root: merkleRoot,
      })));
      throw new Error(`anchor rejected: invalid signature from ${gatewayId}`);
    }

    const key = this.key(ctx, batchId);
    const existing = await ctx.stub.getState(key);
    if (existing.length > 0) {
      throw new Error(`batch ${batchId} already anchored`);
    }

    const anchor: BatchAnchor = {
      batch_id: batchId,
      gateway_id: gatewayId,
      merkle_root: merkleRoot,
      window_start: parseFloat(windowStart),
      window_end: parseFloat(windowEnd),
      reading_count: parseInt(count, 10),
      signature: signatureHex,
      anchored_at: ctx.stub.getTxTimestamp().seconds.toNumber(),
    };
    await ctx.stub.putState(key, Buffer.from(JSON.stringify(anchor)));
    return JSON.stringify(anchor);
  }

  @Transaction(false)
  @Returns("string")
  public async ListAnchors(ctx: Context, gatewayId: string): Promise<string> {
    const iterator = ctx.stub.getStateByPartialCompositeKey("anchor", []);
    const anchors: BatchAnchor[] = [];
    for await (const kv of iterator) {
      const anchor = JSON.parse(kv.value.toString()) as BatchAnchor;
      if (!gatewayId || anchor.gateway_id === gatewayId) {
        anchors.push(anchor);
      }
    }
    anchors.sort((a, b) => a.window_start - b.window_start);
    return JSON.stringify(anchors);
  }

  @Transaction(false)
  @Returns("string")
  public async GetAnchor(ctx: Context, batchId: string): Promise<string> {
    const data = await ctx.stub.getState(this.key(ctx, batchId));
    if (data.length === 0) {
      throw new Error(`no anchor for batch ${batchId}`);
    }
    return data.toString();
  }

  /**
   * Citizen verification: recompute the root from a leaf hash and a
   * Merkle path (JSON array of [siblingHashHex, "left"|"right"]) and
   * compare with the anchored root.
   */
  @Transaction(false)
  @Returns("string")
  public async VerifyReading(ctx: Context, batchId: string, leafHash: string, pathJson: string): Promise<string> {
    const { createHash } = await import("crypto");
    const anchor = JSON.parse(await this.GetAnchor(ctx, batchId)) as BatchAnchor;
    let node = leafHash;
    for (const [sibling, side] of JSON.parse(pathJson) as [string, string][]) {
      const pair = side === "left"
        ? Buffer.concat([Buffer.from([1]), Buffer.from(sibling, "hex"), Buffer.from(node, "hex")])
        : Buffer.concat([Buffer.from([1]), Buffer.from(node, "hex"), Buffer.from(sibling, "hex")]);
      node = createHash("sha256").update(pair).digest("hex");
    }
    const verified = node === anchor.merkle_root;
    return JSON.stringify({
      verified,
      computed_root: node,
      onchain_root: anchor.merkle_root,
      ...(verified ? {} : { reason: "Merkle proof does not match on-chain root — data altered after anchoring" }),
    });
  }
}

File: sensorchain/contracts/anchor.py
"""Merkle-Root Batch Anchor contract.

The gateway agent submits one transaction per batch window containing
the Merkle root of that window's readings, signed by the gateway's HSM
key. The contract verifies the signature against the registered public
key before accepting the anchor — an anchor from an unregistered or
retired gateway is rejected, and a bad signature is rejected and
reported as a tamper event.
"""

from __future__ import annotations

import time

from ..identity import verify_signature
from ..ledger import Ledger
from ..merkle import MerkleProof


def anchor_signing_payload(gateway_id: str, batch_id: str, merkle_root: str,
                           window_start: float, window_end: float, count: int) -> bytes:
    """The exact bytes a gateway signs when anchoring a batch."""
    return f"{gateway_id}|{batch_id}|{merkle_root}|{window_start:.3f}|{window_end:.3f}|{count}".encode("utf-8")


class AnchorContract:
    CONTRACT = "anchor"

    def __init__(self, ledger: Ledger):
        self.ledger = ledger

    @staticmethod
    def _key(batch_id: str) -> str:
        return f"anchor~{batch_id}"

    def anchor_batch(
        self,
        gateway_id: str,
        batch_id: str,
        merkle_root: str,
        window_start: float,
        window_end: float,
        count: int,
        signature_hex: str,
    ) -> dict:
        gateway = self.ledger.get_state(f"device~{gateway_id}")
        if gateway is None:
            raise PermissionError(f"anchor rejected: gateway {gateway_id} is not registered")
        if gateway.get("status") == "retired":
            raise PermissionError(f"anchor rejected: gateway {gateway_id} is retired")

        payload = anchor_signing_payload(gateway_id, batch_id, merkle_root, window_start, window_end, count)
        if not verify_signature(gateway["public_key_pem"], payload, signature_hex):
            self.ledger.emit_event("AnchorSignatureInvalid", {
                "gateway_id": gateway_id, "batch_id": batch_id, "merkle_root": merkle_root,
            })
            raise PermissionError(f"anchor rejected: invalid signature from {gateway_id}")

        if self.ledger.get_state(self._key(batch_id)) is not None:
            raise ValueError(f"batch {batch_id} already anchored")

        anchor = {
            "batch_id": batch_id,
            "gateway_id": gateway_id,
            "merkle_root": merkle_root,
            "window_start": window_start,
            "window_end": window_end,
            "reading_count": count,
            "signature": signature_hex,
            "anchored_at": time.time(),
        }
        tx = self.ledger.put_state(self.CONTRACT, "anchorBatch", self._key(batch_id), anchor)
        anchor["tx_id"] = tx.tx_id
        return anchor

    def get_anchor(self, batch_id: str) -> dict | None:
        return self.ledger.get_state(self._key(batch_id))

    def list_anchors(self, gateway_id: str | None = None) -> list[dict]:
        anchors = [v for _, v in self.ledger.query_by_prefix("anchor~")]
        if gateway_id is not None:
            anchors = [a for a in anchors if a["gateway_id"] == gateway_id]
        return sorted(anchors, key=lambda a: a["window_start"])

    def verify_reading(self, batch_id: str, proof: MerkleProof) -> dict:
        """Citizen-facing verification: does this proof chain up to the
        on-chain root for the claimed batch?"""
        anchor = self.get_anchor(batch_id)
        if anchor is None:
            return {"verified": False, "reason": f"no anchor on-chain for batch {batch_id}"}
        if proof.verify(anchor["merkle_root"]):
            return {"verified": True, "batch_id": batch_id,
                    "merkle_root": anchor["merkle_root"], "anchored_at": anchor["anchored_at"]}
        return {"verified": False, "reason": "Merkle proof does not match on-chain root — data altered after anchoring",
                "expected_root": anchor["merkle_root"], "computed_root": proof.compute_root()}
